pkg/controller/intrusiondetection/intrusiondetection_controller.go

```diff
@@ -88,134 +88,121 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
-		status:          status.New(mgr.GetClient(), "intrusion-detection", opts.KubernetesVersion),
-		clusterDomain:   opts.ClusterDomain,
-		licenseAPIReady: licenseAPIReady,
-		dpiAPIReady:     dpiAPIReady,
-		tierWatchReady:  tierWatchReady,
-		usePSP:          opts.UsePSP,
-		elasticExternal: opts.ElasticExternal,
+	// Determine how to handle watch events for cluster-scoped resources. For multi-tenant clusters,
+	// we should update all tenants whenever one changes. For single-tenant clusters, we can just queue the object.
+	var eventHandler handler.EventHandler = &handler.EnqueueRequestForObject{}
+	if opts.MultiTenant {
+		eventHandler = utils.EnqueueAllTenants(mgr.GetClient())
 	}
-	r.status.Run(opts.ShutdownContext)
-	return r
-}
 
-// add adds watches for resources that are available at startup
-func add(mgr manager.Manager, c controller.Controller) error {
-	var err error
-
-	// Watch for changes to primary resource IntrusionDetection
-	err = c.Watch(&source.Kind{Type: &operatorv1.IntrusionDetection{}}, &handler.EnqueueRequestForObject{})
-	if err != nil {
-		return fmt.Errorf("intrusiondetection-controller failed to watch primary resource: %v", err)
+	policiesToWatch := []types.NamespacedName{
+		{Name: render.IntrusionDetectionControllerPolicyName, Namespace: installNS},
+		{Name: render.IntrusionDetectionInstallerPolicyName, Namespace: installNS},
+		{Name: render.ADAPIPolicyName, Namespace: installNS},
+		{Name: render.ADDetectorPolicyName, Namespace: installNS},
+		{Name: networkpolicy.TigeraComponentDefaultDenyPolicyName, Namespace: installNS},
 	}
-
-	err = c.Watch(&source.Kind{Type: &batchv1.Job{ObjectMeta: metav1.ObjectMeta{
-		Namespace: render.IntrusionDetectionNamespace,
-		Name:      render.IntrusionDetectionInstallerJobName,
-	}}}, &handler.EnqueueRequestForObject{})
-	if err != nil {
-		return fmt.Errorf("intrusiondetection-controller failed to watch installer job: %v", err)
+	if !opts.MultiTenant {
+		// DPI is only supported in single-tenant mode.
+		go utils.WaitToAddResourceWatch(c, k8sClient, log, dpiAPIReady,
+			[]client.Object{&v3.DeepPacketInspection{TypeMeta: metav1.TypeMeta{Kind: v3.KindDeepPacketInspection}}})
+		policiesToWatch = append(policiesToWatch, types.NamespacedName{Name: dpi.DeepPacketInspectionPolicyName, Namespace: dpi.DeepPacketInspectionNamespace})
 	}
+	go utils.WaitToAddNetworkPolicyWatches(c, k8sClient, log, policiesToWatch)
+	go utils.WaitToAddLicenseKeyWatch(c, k8sClient, log, licenseAPIReady)
+	go utils.WaitToAddTierWatch(networkpolicy.TigeraComponentTierName, c, k8sClient, log, tierWatchReady)
 
-	// Watch for changes to to primary resource LogCollector, to determine if syslog forwarding is
-	// turned on or off.
-	err = c.Watch(&source.Kind{Type: &operatorv1.LogCollector{}}, &handler.EnqueueRequestForObject{})
-	if err != nil {
+	// Watch for changes to operator.tigera.io APIs.
+	if err = c.Watch(&source.Kind{Type: &operatorv1.IntrusionDetection{}}, eventHandler); err != nil {
```

Comment:
I think we want this to be `EnqueueRequestForObject` (I remember we talked about this case specifically on my compliance PR) so that we only reconcile the tenant namespace in which the `IntrusionDetection` CR was modified rather than all other tenants.

Reply:
Yep, good catch